src/NSL/parameter.tpp

```diff
@@ -0,0 +1,171 @@
+ *        such that it can be used together with `NSL::Parameter::operator[]`;
+ * */
+class ParameterEntry{
+    public:
+    ParameterEntry() = default;
+
+    //! Retrieve the stored value. Type must match the initial type otherwise runtime error is thrown.
+    template<typename Type>
+    Type & to(){
+        // get the derived TemplatedParameterEntry
+        TemplatedParameterEntry<Type> * ptr = upcast<Type>();
+        return ptr->value_;
+    }
+
+    //! Retrieve the stored value. Type must match the initial type otherwise runtime error is thrown.
+    template<typename Type>
+    const Type & to() const {
+        // get the derived TemplatedParameterEntry
+        TemplatedParameterEntry<Type> * ptr = upcast<Type>();
+        return ptr->value_;
+    }
+
+    //! Implicitly convert to Type. Type must match the initial type otherwise runtime error is thrown.
+    template<typename Type>
+    operator Type(){
+        // get the derived TemplatedParameterEntry
+        TemplatedParameterEntry<Type> * ptr = upcast<Type>();
+        return ptr->value_;
+    }
+
+    //! Implicitly convert to Type. Type must match the initial type otherwise runtime error is thrown.
+    template<typename Type>
+    operator Type() const {
+        // get the derived TemplatedParameterEntry
+        TemplatedParameterEntry<Type> * ptr = upcast<Type>();
+        return ptr->value_;
+    }
+
+    //! Assigns a value of Type. Type must match the initial type otherwise runtime error is thrown.
+    template<typename Type>
+    ParameterEntry operator=(const Type & value){
+        // get the derived TemplatedParameterEntry
+        TemplatedParameterEntry<Type> * ptr = upcast<Type>();
+        ptr->value_ = value;
+        return *ptr;
+    }
+
+    virtual ~ParameterEntry() = default;
+
+    protected:
+    //! Retrieves the address of `NSL::TemplatedParameterEntry` the runtime polymorphism points to
+    /*!
+     * This function performs an upcast to the derived onject `template<typename Type> NSL::TemplateParameterentry`.
+     * and performs checks that the conversion worked out correctly.
+     * If dynamic_cast fails (returning a nullptr) a std::runtime_error is thrown
+     * */
+    template<typename Type>
+    NSL::TemplatedParameterEntry<Type> * upcast(){
+        // if conversion does not work ptr = nullptr or throw std::bad_cast
+        NSL::TemplatedParameterEntry<Type> * ptr = dynamic_cast<TemplatedParameterEntry<Type> *>(this);
+        
+        // check for nullptr and return or throw
+        if(ptr){
+            return ptr;
+        } else {
+            throw std::runtime_error("Converting ParameterEntry to TemplatedParameterEntry failed");
+        }
+    }
+};
+
+/*! 
+ * This class implements a wrapper around a simple data type `Type` and is
+ * implementing the back end for `NSL::ParameterEntry`.
+ * The user most likely should not need to worry about this class as the point
```

Comment:
Should we put it in a namespace that would be hidden to the typical user?

Reply:
I tried that but it didn't compile. Therefore, I simply went back and put a comment.

Reply:
Resolved by opening #148.